src/Logary/TargetModule.fs

```diff
@@ -26,95 +28,45 @@ module Target =
     member x.accepts message =
       Rule.accepts message x.rules
 
-  /// WARNING: this may cause you headaches; are you sure you don't want to use tryLog?
-  ///
-  /// Returns an Alt that commits on the Targets' buffer accepting the message.
-  let log (x: T) (msg: Message): LogResult =
-    let msg = x.transform msg
-    if not (x.accepts msg) then LogResult.rejected
-    else
-      let ack = IVar ()
-      let targetMsg = Log (msg, ack)
-      RingBuffer.put x.api.requests targetMsg ^->.
-      Ok (upcast ack)
 
-  /// If the alternative is committed to, the RingBuffer WILL enqueue the message. The tryLog operation is done as
-  /// `Ch.give x.tryPut m`.
+  /// Returns an Alt that commits on the Targets' buffer accepting the message.
   ///
-  /// Returns an Alt that commits on the Targets' buffer accepting the message OR the RingBuffer telling the caller that
-  /// it was full, and returning an Error(BufferFull [targetName]) LogResult.
-  let tryLog (x: T) (msg: Message): LogResult =
+  /// - `putBufferTimeOut` means it will try to wait this duration for putting the message onto buffer,otherwise return timeout error, and cancel putting message.
+  ///   if `putBufferTimeOut <=  Duration.Zero` , it will try to detect whether the buffer is full and return immediately
+  let log (putBufferTimeOut: Duration) (x: T) (msg: Message): LogResult =
     let msg = x.transform msg
-    if not (x.accepts msg) then LogResult.rejected
+    if not (x.accepts msg) then LogResult.notAcceptByTarget x.name
     else
       let ack = IVar ()
       let targetMsg = Log (msg, ack)
-      // This returns immediately if the buffer is full; and the RingBuffer's server is
-      // always available to notify the caller that the buffer is full.
-      RingBuffer.tryPut x.api.requests targetMsg ^-> function
+      if putBufferTimeOut <= Duration.Zero then
+        RingBuffer.tryPut x.api.requests targetMsg ^-> function
         | true ->
           Result.Ok (upcast ack)
         | false ->
-          Result.Error (BufferFull x.name)
+          LogError.targetBufferFull x.name
+      else
+        RingBuffer.put x.api.requests targetMsg ^->. Result.Ok (upcast ack)
+        <|>
+        timeOut (putBufferTimeOut.ToTimeSpan()) ^-> fun _ -> LogError.timeOutToPutBuffer x.name putBufferTimeOut.TotalSeconds
```

Comment:
This is what caused our cascading failures of logging high-throughput at tradera.com. We can't go back to it.

Reply:
do we have some perf-test to check this situation?  like this one ? #322

Reply:
Yes, let's benchmark it.